### ISSUE TITLE
[r-apt] `"installBspm": true` is not working on Debian

### ISSUE THREAD

**Issue by eitsupi** [rocker-org/devcontainer-features#173]
I have confirmed that `install.packages("curl")` fails on R started by a non-root user (`vscode`) on an image built from devcontainer.json as shown below.

```json
{
	"image": "mcr.microsoft.com/devcontainers/base:debian",
	"features": {
		"ghcr.io/rocker-org/devcontainer-features/r-apt:0": {
			"installBspm": true
		}
	}
}
```

![image](https://github.com/rocker-org/devcontainer-features/assets/50911393/8304ee9a-15c9-46f4-917a-07b6f36c2c1a)

This setup corresponds to the following script run with `INSTALLBSPM=true`.
https://github.com/rocker-org/devcontainer-features/blob/583387a2831c181905c571b306f3a976affbe5cb/src/r-apt/install.sh

Perhaps the script contains improper settings somewhere?

_Originally posted by @eitsupi in https://github.com/rocker-org/devcontainer-features/issues/169#issuecomment-1671237933_

**Comment by Enchufa2** [comment]
It does update the cache here:

https://github.com/Enchufa2/bspm/blob/4d3c8ce500c0c4f7b30ca322fdf51d9c26b23d9d/inst/service/backend/apt.py#L15

But note that the error is raised one line before that, when we are just instantiating the `Cache` object. So I suppose that `apt update` does _something else_ on top of that, maybe to fix inconsistent states?

bspm runs just fine with `debian:testing`, but the base image you are using here is a stable one, right? Then I'd say it's your responsibility to ensure that the cache is in a consistent state after switching to `testing`.

**Comment by Enchufa2** [comment]
There's definitely something wrong with that setting. I see:

```bash
$ docker run --rm -it mcr.microsoft.com/devcontainers/base:debian
root ➜ / $ echo 'APT::Default-Release "testing";' >/etc/apt/apt.conf.d/default
root ➜ / $ apt-get update
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8906 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [4732 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [51.9 kB]
Fetched 9214 kB in 1s (10.4 MB/s)                     
Reading package lists... Done
root ➜ / $ apt-get install r-base
Reading package lists... Done
E: The value 'testing' is invalid for APT::Default-Release as such a release is not available in the sources
```

Maybe @eddelbuettel knows more about this setting.

**Comment by eitsupi** [comment]
Thanks,

Interestingly, a search on that error message turned up previous issues of the Rocker project. (rocker-org/rocker#167)

**Comment by Enchufa2** [comment]
It's strange, because, according to the docs,

```python
import apt

# First of all, open the cache
cache = apt.Cache()
# Now, lets update the package list
cache.update()
```

should be equivalent to `apt-get update`, which succeeds. But opening the cache fails in the first place. Maybe this is a bug in the Python API?

**Comment by Enchufa2** [comment]
Another question is: do you _really_ need to set `APT::Default-Release "testing";` in the first place? If I spawn a `debian:stable` container and add the `testing` repo, package installations come from `testing`, and this bug is no more.

**Comment by Enchufa2** [comment]
All package managers install the latest package versions by default, and APT is no exception. So if you just add the testing repositories, the latest R will be pulled.

From what I've seen, the typical use of `APT::Default-Release` would be the other way around: to pin packages by default to an **older** release when you add repos from a newer one. I.e. I want `r-base` from testing, but I don't want the rest of the packages to be automatically upgraded to testing, so I add the testing repos but set `APT::Default-Release` to `stable`.

TL;DR, unless I'm missing something (please @eddelbuettel correct me if I'm wrong), I think you can just drop that line and everything will work as intended here.

**Comment by eitsupi** [comment]
After #204 merging, it seems working now.
As @Enchufa2 pointed out, it seems that `APT::Default-Release "testing";` was the cause. thank you!
